### PR TITLE
Remove the .data section from RISC-V's default linker script addon

### PR DESCRIPTION
**Pull request by palmer-dabbelt**
I'm actually not sure why this is here: .data isn't read-only and
shouldn't be in the INITIAL_READONLY_SECTIONS variable.  This causes a
handful of sections that should be read-only to actually be read-write
when loaded, which breaks things like "-z relro".

Removing this causes the default .data mapping to be used, which comes
from upstream.  As far os I can tell, there was no reason we were ever
supposed to have our own .data in the first place, and no other
architectures have ever done it this way.

I tested this still builds and boots Busybox on Linux.

Thanks to Alan Morda for finding the problem.

ld/ChangeLog

2016-12-29  Palmer Dabbelt <palmer@dabbelt.com>

        * emulparams/elf32lriscv-defs.sh (INITIAL_READONLY_SECTIONS):
        Remove the RISC-V specific .data section mapping.

### CODE REVIEW

File: ld/emulparams/elf32lriscv-defs.sh [hunk 1]
```diff
@@ -29,7 +29,6 @@ SDATA_START_SYMBOLS="_gp = . + 0x800;
 # Place the data section before text section.  This enables more compact
 # global variable access for RVC code via linker relaxation.
 INITIAL_READONLY_SECTIONS="
-  .data           : { *(.data) *(.data.*) *(.gnu.linkonce.d.*) }
   .rodata         : { *(.rodata) *(.rodata.*) *(.gnu.linkonce.r.*) }
   .srodata        : { ${SDATA_START_SYMBOLS} }
   .sdata          : { *(.sdata .sdata.* .gnu.linkonce.s.*) }
```
Comment (kito-cheng) [inline ld/emulparams/elf32lriscv-defs.sh]: sdata mean small data, so I think you need remove it too.

**Comment by aswaterman** [comment]
Kito's right.  But I suggest just deleting the entire INITIAL_READONLY_SECTIONS statement.  It is a micro-optimization for RVC that adds to the "weirdness budget" but does not help much.  I did that in this commit: https://github.com/riscv/riscv-gnu-toolchain/commit/97951e32a38043c7f03db08178313a14b13dc5a1

I'll force-push over this branch with a partial revert of that commit.  (We don't want to move the text section again--it's fine at 0x10000.)

**Comment by palmer-dabbelt** [comment]
@aswaterman I just pushed Kito's suggestion, IDK if I overwrote whatever you were doing.  Can you just make this what you want and then ping me?

**Comment by palmer-dabbelt** [comment]
@aswaterman It looks like your patch breaks the glibc build:

````
mv -f /scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/libc.so.6.new /scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/libc.so.6
riscv64-unknown-linux-gnu-gcc -march=rv32ima -mabi=ilp32 -nostdlib -nostartfiles -o /scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/iconv/iconvconfig    -Wl,-z,combreloc -Wl,-z,relro -Wl,--hash-style=both /scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/csu/crt1.o /scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/csu/crti.o `riscv64-unknown-linux-gnu-gcc -march=rv32ima -mabi=ilp32  --print-file-name=crtbegin.o` /scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/iconv/iconvconfig.o /scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/iconv/strtab.o /scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/iconv/xmalloc.o /scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/iconv/hash-string.o  -Wl,-dynamic-linker=/lib32/ilp32/ld.so.1 -Wl,-rpath-link=/scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32:/scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/math:/scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/elf:/scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/dlfcn:/scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/nss:/scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/nis:/scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/rt:/scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/resolv:/scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/crypt:/scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/mathvec:/scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/nptl /scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/libc.so.6 /scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/libc_nonshared.a -Wl,--as-needed /scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/elf/ld.so -Wl,--no-as-needed -lgcc  `riscv64-unknown-linux-gnu-gcc -march=rv32ima -mabi=ilp32  --print-file-name=crtend.o` /scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/csu/crtn.o
/scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/build-linux-multilib/build-glibc-linux-rv32ima-ilp32/csu/crt1.o: In function `.L0 ':
/scratch/palmer.dabbelt/work/upstream/riscv-gnu-toolchain/riscv-glibc/csu/../sysdeps/riscv/start.S:58: undefined reference to `_gp'
````

Do we need to keep the SDATA_START_SYMBOLS line?  That seems to be where we get _gp from.

**Comment by aswaterman** [comment]
D'oh.  Let me look...

**Comment by aswaterman** [comment]
Yes, but that's not the place to put it.  Let me see how others do it...

**Comment by aswaterman** [comment]
attempt 2 is pushed (fortunately it deleted yet more code). can you retry?

**Comment by palmer-dabbelt** [comment]
GCC stage 2 is building, so it's probably fine.

**Comment by palmer-dabbelt** [comment]
I guess it ended up building :).

**Pull request closed without merging**